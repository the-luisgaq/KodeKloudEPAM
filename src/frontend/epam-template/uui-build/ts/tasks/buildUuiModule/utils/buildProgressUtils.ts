import { logger } from '../../../utils/jsBridge';
import { readJsonFileSync } from '../../../utils/fileUtils';
import path from 'path';

type TBuildProgressParams = {
    moduleName: string,
    isRollup: boolean
};
class ModuleBuildProgressLogger {
    private startMs: number | undefined;

    constructor(public params: TBuildProgressParams) {}

    start() {
        this.initTimer();
        const notes = this.params.isRollup ? '(rollup)' : '(no index.tsx; will be published as-is)';
        logger.info(`${this.getMsgHeader()} ${notes}`);
    }

    done() {
        logger.success(`${this.getMsgHeader()} Done (${this.getTimerSinceStart()})`);
    }

    error() {
        logger.error(`${this.getMsgHeader()} Failed`);
    }

    private getMsgHeader() {
        return `Build "${this.params.moduleName}"`;
    }

    private initTimer() {
        this.startMs = Date.now();
    }

    private getTimerSinceStart(): string {
        const ms = Date.now() - this.startMs!;
        return `took ${Number(ms / 1000).toFixed(2)} sec`;
    }
}

export async function withEventsLogger(
    params: { moduleRootDir: string, isRollup: boolean, asyncCallback: () => Promise<void> },
) {
    const { asyncCallback, isRollup, moduleRootDir } = params;
    const moduleName = getModuleName(moduleRootDir);
    const moduleBuildLogger = new ModuleBuildProgressLogger({ moduleName, isRollup });
    moduleBuildLogger.start();
    try {
        await asyncCallback();
        moduleBuildLogger.done();
    } catch (err) {
        moduleBuildLogger.error();
        err && err.message && logger.error(err.message);
        process.exit(1);
    }
}

function getModuleName(moduleRootDir: string) {
    const res = readJsonFileSync(path.resolve(moduleRootDir, 'package.json'));
    if (res) {
        return `${res.name}@${res.version}`;
    }
    return moduleRootDir;
}
