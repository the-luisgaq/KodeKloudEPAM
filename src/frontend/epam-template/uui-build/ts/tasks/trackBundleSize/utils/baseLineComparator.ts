import { COMPARISON_THRESHOLD_PERCENTAGE } from '../constants';
import { TBundleSize, TBundleSizeBaseLine, TComparisonRes } from '../types';

/**
 * It expects all sizes specified in bytes, but the Markdown output will be generated in kBytes (for readability).
 */
export function compareBaseLines(params: { currentBaseLine: TBundleSizeBaseLine | undefined, newBaseLine: TBundleSizeBaseLine }): TComparisonRes {
    const { currentBaseLine, newBaseLine } = params;
    return Object.keys(newBaseLine.sizes).reduce<TComparisonRes>((acc, name) => {
        const size = newBaseLine.sizes[name];
        const sizeTotal = size.js + size.css;
        /**
         * We use 0 as a baseline for any new packages without previous baseline
         */
        const baseLineSize: TBundleSize = currentBaseLine?.sizes[name] || { css: 0, js: 0 };
        const threshold = getThreshold(baseLineSize);
        const diff = getDiff(baseLineSize, size);
        const diffSeparated = getDiffSeparated(baseLineSize, size);
        const withinThreshold = sizeTotal >= threshold[0] && sizeTotal <= threshold[1];

        const diffDetails = getDiffDetailsMd(diffSeparated);
        acc[name] = {
            baseLineSize: bytesToKbSeparated(baseLineSize),
            size: bytesToKbSeparated(size),
            diffLabel: `${getSign(diff)}${bytesToKb(diff)}<br />${diffDetails}`,
            withinThreshold,
            thresholdLabel: `${bytesToKb(threshold[0])} - ${bytesToKb(threshold[1])}`,
        };
        return acc;
    }, {});
}

function getDiffDetailsMd(params:{ js: number, css: number }) {
    const js = `${getSign(params.js)}${bytesToKb(params.js)}`;
    const css = `${getSign(params.css)}${bytesToKb(params.css)}`;
    const fence = '```';
    return `${fence} js:${js}${fence}<br />${fence}css:${css} ${fence}`;
}

function getSign(diff: number) {
    return diff > 0 ? '+' : '';
}

function normalizeSizeNumber(num: number) {
    return Number(Number(num).toFixed(2));
}
function bytesToKbSeparated(bytes: TBundleSize) {
    return {
        js: bytesToKb(bytes.js),
        css: bytesToKb(bytes.css),
    };
}

function bytesToKb(bytes: number) {
    return normalizeSizeNumber(bytes / 1024);
}

function getThreshold(size: TBundleSize) {
    const total = size.css + size.js;
    const pcNorm = COMPARISON_THRESHOLD_PERCENTAGE / 100;
    return [Math.floor(total * (1 - pcNorm)), Math.ceil(total * (1 + pcNorm))];
}

function getDiff(prevSize: TBundleSize, nextSize: TBundleSize) {
    return (nextSize.js + nextSize.css) - (prevSize.js + prevSize.css);
}
function getDiffSeparated(prevSize: TBundleSize, nextSize: TBundleSize): TBundleSize {
    return {
        js: nextSize.js - prevSize.js,
        css: nextSize.css - prevSize.css,
    };
}
