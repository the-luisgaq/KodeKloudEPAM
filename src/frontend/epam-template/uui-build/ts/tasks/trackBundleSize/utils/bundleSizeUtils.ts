import path from 'path';
import SourceMapExplorer from 'source-map-explorer';
import { TBundleSize } from '../types';
import { ExploreResult } from 'source-map-explorer/lib/types';
import { uuiRoot } from '../../../constants';

export async function getBundleSizeByRelGlob(relativeGlob: { css: string, js: string }): Promise<TBundleSize> {
    const absGlob = {
        css: path.resolve(uuiRoot, relativeGlob.css),
        js: path.resolve(uuiRoot, relativeGlob.js),
    };
    return await getBundleSizeByAbsGlob(absGlob);
}
export async function getBundleSizeByAbsGlob(absoluteGlob: { css: string, js: string }): Promise<TBundleSize> {
    const [css, js] = await Promise.all([
        getSingleSizeByAbsGlob(absoluteGlob.css),
        getSingleSizeByAbsGlob(absoluteGlob.js),
    ]);
    return { css, js };
}
async function getSingleSizeByAbsGlob(absGlob: string): Promise<number> {
    const result = await explore(absGlob);
    return result.bundles.reduce((acc, { totalBytes }) => {
        return acc + totalBytes;
    }, 0);
}

async function explore(absGlob: string) {
    try {
        return await SourceMapExplorer(absGlob, { output: { format: 'json' }, noBorderChecks: true, gzip: false });
    } catch (err: any) {
        if (err.bundles?.length === 0 && err.errors?.length === 1 && err.errors[0].code === 'ENOENT') {
            return err as ExploreResult;
        } else {
            console.error(err);
            console.error('Unable to measure size of: ' + absGlob);
            throw new Error(err);
        }
    }
}
