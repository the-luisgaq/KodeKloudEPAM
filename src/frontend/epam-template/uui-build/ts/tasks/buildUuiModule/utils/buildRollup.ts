import rollup, { OutputOptions, RollupBuild, RollupOptions } from 'rollup';
import { withEventsLogger } from './buildProgressUtils';
import { getIndexFileRelativePath } from '../../../utils/jsBridge';
// @ts-ignore
import { createRollupConfigForModule } from '../../../../rollup/rollup.config.js';

interface IRollupBuildParams {
    /**
     * "absolute" path to the module root dir
     */
    moduleRootDir: string;
    /**
     * array of file/folder names in the module root dir
     * which must be copied to the output "as-is" (except for package.json which is always copied anyway).
     */
    copyAsIs: string[];
    /**
     * pass a callback if you need to override default behavior for external modules
     */
    external?: (params: { moduleRootDir: string }) => string[];
    /**
     * callback if content of copied package.json needs to be adjusted
     */
    packageJsonTransform: (content: Record<string, unknown>) => object;
}

/**
 * Builds module using Rollup.
 */
export async function buildModuleUsingRollup(options: IRollupBuildParams): Promise<void> {
    const { moduleRootDir } = options;
    const asyncCallback = async () => {
        await buildUsingRollup(options);
    };
    await withEventsLogger({ moduleRootDir, isRollup: true, asyncCallback });
}

/**
 * Starts Rollup build for the given module
 */
async function buildUsingRollup(params: IRollupBuildParams): Promise<void> {
    const cfg = await getConfigEffective(params);
    const { output: outputConfig, ...inputConfig } = cfg[0];
    let bundle: RollupBuild;
    const cleanup = async () => {
        bundle && (await bundle.close());
    };
    try {
        bundle = await rollup.rollup({ ...inputConfig });
        await Promise.all((outputConfig as OutputOptions[]).map(bundle.write));
    } catch (error) {
        console.error(error);
        process.exit(1);
    } finally {
        await cleanup();
    }
}

/**
 * Creates rollup config for the module
 */
async function getConfigEffective(params: IRollupBuildParams): Promise<RollupOptions[]> {
    const indexFileRelativePath = await getIndexFileRelativePath(params.moduleRootDir);
    return await createRollupConfigForModule({
        ...params,
        isWatch: false,
        indexFileRelativePath,
    });
}
