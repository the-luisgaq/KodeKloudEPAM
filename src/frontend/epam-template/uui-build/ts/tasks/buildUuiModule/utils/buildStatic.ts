// @ts-ignore
import fs from 'fs-extra';
import path from 'path';
import { FOLDERS } from '../constants';
import { withEventsLogger } from './buildProgressUtils';

/**
 * This is to handle modules which don't require build.
 * This function just copies all files to the output dir.
 */
export async function buildStaticModule(params: { moduleRootDir: string }) {
    const { moduleRootDir } = params;
    const asyncCallback = async () => {
        fs.emptyDirSync(FOLDERS.build);
        copyAllModuleFilesToOutputSync(moduleRootDir);
    };
    await withEventsLogger({ moduleRootDir, isRollup: false, asyncCallback });
}

/**
 * Copy everything to the output folder "as-is".
 */
function copyAllModuleFilesToOutputSync(moduleRootDir: string) {
    for (const file of fs.readdirSync(moduleRootDir)) {
        if ([FOLDERS.build, FOLDERS.node_modules].indexOf(file) === -1) {
            const from = path.resolve(moduleRootDir, file);
            const to = path.resolve(moduleRootDir, FOLDERS.build, file);
            fs.copySync(from, to);
        }
    }
}
