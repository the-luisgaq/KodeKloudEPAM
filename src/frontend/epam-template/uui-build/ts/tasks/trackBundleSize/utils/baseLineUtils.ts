import path from 'path';
import fs from 'fs';
import { BASE_LINE_PATH, TRACK_BUNDLE_SIZE_REPORT_MD } from '../constants';
import { getUuiVersion, logger } from '../../../utils/jsBridge';
import { TBundleSizeBaseLine, TBundleSizeMap } from '../types';
import { uuiRoot } from '../../../constants';
import { createFileSync, readJsonFileSync } from '../../../utils/fileUtils';

const pathToBaselineResolved = path.resolve(uuiRoot, BASE_LINE_PATH);

export function createBaseLine(sizes: TBundleSizeMap): TBundleSizeBaseLine {
    const timestamp = new Date().toISOString().split('T')[0];
    const uuiVersion = getUuiVersion();
    return {
        version: uuiVersion,
        timestamp,
        sizes,
    };
}

export const baseLineToString = (baseline: TBundleSizeBaseLine) => {
    return JSON.stringify(baseline, undefined, 2);
};

/**
 * Creates new file if it doesn't exist.
 * @param baselineStr
 */
export function overrideBaseLineFileSync(baselineStr: string) {
    createFileSync(pathToBaselineResolved, baselineStr);
    logger.info(`New baseline generated at: "${pathToBaselineResolved}".`);
}

export function getCurrentBaseLineSync() {
    if (fs.existsSync(pathToBaselineResolved)) {
        const data = readJsonFileSync(pathToBaselineResolved);
        return data as TBundleSizeBaseLine;
    }
}

export function saveComparisonResultsMd(contentMd: string) {
    const crPathResolved = path.resolve(uuiRoot, TRACK_BUNDLE_SIZE_REPORT_MD);
    createFileSync(crPathResolved, contentMd);
    logger.info(`Comparison result generated at: "${crPathResolved}".`);
}
