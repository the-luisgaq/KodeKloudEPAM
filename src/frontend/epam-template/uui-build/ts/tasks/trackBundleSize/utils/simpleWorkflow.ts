import { logger } from '../../../utils/jsBridge';

export type TSimpleWorkflowStep<P> = ((params: P) => Promise<void>) | (() => Promise<void>);

export function createSimpleWorkflow<P>(map: Record<string, TSimpleWorkflowStep<P>>) {
    async function run(params: P) {
        const mainTimer = getSimpleTimer('main');
        mainTimer.start();
        const keys = Object.keys(map);
        for (let i = 0; i < keys.length; i++) {
            logger.info('');
            const name = keys[i];
            const f = map[name];
            const timer = getSimpleTimer(name);
            timer.start();
            await f(params);
            timer.end();
        }
        mainTimer.end();
    }

    return { run };
}

function getSimpleTimer(name: string) {
    let start: Date;
    let end: Date;

    return {
        start: () => {
            start = new Date();
            logger.info(`Started "${name}".`);
        },

        end: () => {
            end = new Date();
            const sec = Math.ceil((end.getTime() - start.getTime()) / 1000);
            logger.success(`Completed "${name}". Took: ${sec} (sec).`);
        },

    };
}
