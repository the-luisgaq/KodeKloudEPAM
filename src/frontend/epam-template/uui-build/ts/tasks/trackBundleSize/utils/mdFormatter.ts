import { TBundleSizeBaseLine, TComparisonRes, TComparisonResLine } from '../types';
import { baseLineToString } from './baseLineUtils';

export function convertComparisonResultToMd(
    params: {
        currentBaseLine: TBundleSizeBaseLine | undefined,
        newBaseLine: TBundleSizeBaseLine
        comparisonResult: TComparisonRes,
    },
) {
    const { currentBaseLine, newBaseLine, comparisonResult } = params;
    const isExceedsLimits = Object.keys(comparisonResult).reduce((acc, rowId) => {
        if (!comparisonResult[rowId].withinThreshold) {
            return true;
        }
        return acc;
    }, false);

    const baselineVersion = currentBaseLine ? `v${currentBaseLine.version}` : 'baseline not found';
    const headerRemap = {
        baseLineSize: `Baseline Size<br />(${baselineVersion})`,
        size: 'Size',
        diffLabel: 'Diff',
        withinThreshold: 'Within<br />Threshold',
        thresholdLabel: 'Threshold<br />(min - max)',
    };
    const formatHeader = (h: keyof TComparisonResLine) => {
        return headerRemap[h] || h;
    };

    const tableMd = formatMdTable(
        comparisonResult,
        formatHeader,
        (h, v: any) => {
            if (h === 'withinThreshold') {
                return v ? ':ok:' : ':no_entry:';
            }
            if (h === 'baseLineSize' || h === 'size') {
                return String(normalizeSizeNumber(v.js + v.css));
            }
            return v;
        },
    );
    /*
     * The "CI Status: error" part is used by "qualityGate.yml" to mark the workflow failed when limits are exceeded.
     * So, if you want to change it, then please make sure that "qualityGate.yml" is updated as well.
     */
    const generatedBy = 'Generated by: track-bundle-size';
    const ciStatus = `CI Status: ${isExceedsLimits ? 'error' : 'ok'}`;
    const descriptionMd = [
        generatedBy,
        `Generated at: ${new Date().toUTCString()}`,
        'Bundle size diff (in kBytes). Not gzipped. Both CSS & JS included.',
        `Baseline: ${baselineVersion} ${currentBaseLine ? `(${currentBaseLine.timestamp})` : ''}`,
        `${ciStatus}\n`,
    ].join('<br>');

    const newBaseLineStr = baseLineToString(newBaseLine);
    const newBaseLineMd = formatNewBaseLine(newBaseLineStr);

    return `${descriptionMd}\n${tableMd}\n${newBaseLineMd}`;
}

function formatMdTable(
    obj: TComparisonRes,
    formatHeader: (h: keyof TComparisonResLine) => string,
    formatValue: (h: keyof TComparisonResLine, v: TComparisonResLine[keyof TComparisonResLine]) => string,
) {
    const attrs: (keyof TComparisonResLine)[] = ['baseLineSize', 'size', 'diffLabel', 'withinThreshold', 'thresholdLabel'];
    const header = '| Module |' + attrs.map(formatHeader).join('|') + '|';
    const headerSep = '|:-----:|' + attrs.map(() => ':-----:').join('|') + '|';
    const rows = Object.keys(obj).reduce<string[]>((acc, rowId) => {
        const rowContent = attrs.map((a) => {
            return formatValue(a, obj[rowId][a]);
        }).join('|');
        acc.push(`|${rowId}|${rowContent}|`);
        return acc;
    }, []);
    return `${header}\n${headerSep}\n${rows.join('\n')}`;
}

function formatNewBaseLine(baseLine: string) {
    return [
        '<details>',
        '<summary>new sizes (raw)</summary>',
        '',
        'To set the sizes as a new baseline, you can copy/paste next content to the ```uui-build/config/bundleSizeBaseLine.json``` and commit the file.',
        '',
        '```json',
        baseLine,
        '```',
        '</details>',
    ].join('\n');
}

function normalizeSizeNumber(num: number) {
    return Number(Number(num).toFixed(2));
}
