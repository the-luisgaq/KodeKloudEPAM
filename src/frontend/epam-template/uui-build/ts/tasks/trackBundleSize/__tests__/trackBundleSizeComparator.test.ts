import { compareBaseLines } from '../utils/baseLineComparator';

const header = { version: 'test', timestamp: 'test' };
const TestData = {
    currentBaseLine: { ...header, sizes: { testApp1: { js: 70, css: 30 }, testApp2: { js: 100, css: 100 } } },
    newBaseLine: { ...header, sizes: { testApp1: { js: 80, css: 40 }, testApp2: { js: 100, css: 110 } } },
};

describe('trackBundleSizeComparator', () => {
    it('should create comparison result report', () => {
        const result = compareBaseLines({ currentBaseLine: TestData.currentBaseLine, newBaseLine: TestData.newBaseLine });
        expect(result).toMatchSnapshot();
    });
});
