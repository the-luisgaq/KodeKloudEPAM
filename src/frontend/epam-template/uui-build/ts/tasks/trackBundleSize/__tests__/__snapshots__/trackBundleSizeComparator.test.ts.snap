// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`trackBundleSizeComparator should create comparison result report 1`] = `
Object {
  "testApp1": Object {
    "baseLineSize": Object {
      "css": 0.03,
      "js": 0.07,
    },
    "diffLabel": "+0.02<br />\`\`\` js:+0.01\`\`\`<br />\`\`\`css:+0.01 \`\`\`",
    "size": Object {
      "css": 0.04,
      "js": 0.08,
    },
    "thresholdLabel": "0.09 - 0.11",
    "withinThreshold": false,
  },
  "testApp2": Object {
    "baseLineSize": Object {
      "css": 0.1,
      "js": 0.1,
    },
    "diffLabel": "+0.01<br />\`\`\` js:0\`\`\`<br />\`\`\`css:+0.01 \`\`\`",
    "size": Object {
      "css": 0.11,
      "js": 0.1,
    },
    "thresholdLabel": "0.18 - 0.22",
    "withinThreshold": true,
  },
}
`;
